### PR TITLE
Change ColumnAssociationResolver to a Visitor and Move Out of specs.py

### PR DESCRIPTION
**Pull request by plypaul**
### Description

As a part of splitting `specs.py`, this PR moves `ColumnAssociationResolver` out of the file. To resolve circular dependencies, it changes `ColumnAssociationResovler` to use `InstanceSpecVisitor`.

<!--- 
  Before requesting review, please make sure you have:
  1. read [the contributing guide](https://github.com/dbt-labs/metricflow/blob/main/CONTRIBUTING.md),
  2. signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
  3. run `changie new` to [create a changelog entry](https://github.com/dbt-labs/metricflow/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: metricflow/plan_conversion/column_resolver.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ def visit_metadata_spec(self, metadata_spec: MetadataSpec) -> ColumnAssociation:
         )
 
 
-class DefaultColumnAssociationResolver(ColumnAssociationResolver):
+class DunderColumnAssociationResolver(ColumnAssociationResolver):
```
Comment (tlento) [inline metricflow/plan_conversion/column_resolver.py:92]: Oooh I like the rename....

**Review by tlento — APPROVED** [review]
This is fine, I guess, but why do we need yet another visitor here? This does exactly one thing. What's coming in the future that justifies a wide open `accept(ThingThatDoesStuff) -> ArbitraryOutput` API on these instance objects? 

I didn't love the old API, but it seems like it would've worked just fine with a simple move and rename of the resolver itself.

File: metricflow/plan_conversion/column_resolver.py [hunk 2]
```diff
@@ -65,7 +65,12 @@ def resolve_time_dimension_spec(  # noqa: D
             ).qualified_name
 
         return ColumnAssociation(
-            column_name=column_name + (f"__{aggregation_state.value.lower()}" if aggregation_state else ""),
+            column_name=column_name
+            + (
+                f"__{time_dimension_spec.aggregation_state.value.lower()}"
```
[the remark this answers is not in the document]
Comment (plypaul) [reply metricflow/plan_conversion/column_resolver.py:69]: > Is this a dunder? Should we make it a constant?

Yeah, changed to use the defined constant.

> Also we should probably figure out what to do with time granularity adjustments if we are not using dunder in the future.

💯

**Comment by plypaul** [comment]
> I didn't love the old API, but it seems like it would've worked just fine with a simple move and rename of the resolver itself.

There are some later PRs that become easier to write with that interface. Since I haven't pushed those out yet, there's still possibility to revert if it does not prove useful.

**Pull request merged**